font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,157 @@
+            },
+            Err(err) => Err(err),
+        }
     }
 
-    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
-        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+    fn get_fallback_font(&self, font: &Font, c: char) -> Result<Font, Error> {
+        let fallback = match self.font_fallback.as_ref() {
+            Some(fallback) => fallback,
+            None => return Err(Error::MissingGlyph(c)),
+        };
+
+        let utf16_codepoints = encode_utf16(c);
+        let length = utf16_codepoints.len() as u32;
+
+        let locale = get_current_locale();
+        let text_analysis_source = dwrote::TextAnalysisSource::from_text(
+            Box::new(TextAnalysisSourceData { locale, length }),
+            utf16_codepoints,
+        );
+
+        let fallback_result = fallback.map_characters(
+            &text_analysis_source,
+            0,
+            length,
+            &self.font_collection,
+            Some(&font.family_name()),
+            font.weight(),
+            font.style(),
+            font.stretch(),
+        );
+
+        fallback_result.mapped_font.ok_or_else(|| Error::MissingGlyph(c))
+    }
+
+    fn rasterize_font_char(
+        &self,
+        font_face: &FontFace,
+        c: char,
+        size: Size,
+    ) -> Result<RasterizedGlyph, Error> {
+        let offset = GlyphOffset { advanceOffset: 0.0, ascenderOffset: 0.0 };
+
+        let glyph_index = *font_face
+            .get_glyph_indices(&[c as u32])
+            .first()
+            .ok_or_else(|| Error::MissingGlyph(c))?;
+
+        // The DirectWrite documentation states that we should get 0 returned if the glyph
+        // does not exist in the font
+        if glyph_index == 0 {
+            return Err(Error::MissingGlyph(c));
+        }
+
+        let em_size = em_size(size);
+
+        let glyph_run = dwrote::DWRITE_GLYPH_RUN {
+            fontFace: unsafe { font_face.as_ptr() },
+            fontEmSize: em_size,
+            glyphCount: 1,
+            glyphIndices: &(glyph_index),
+            glyphAdvances: &(0.0),
+            glyphOffsets: &(offset),
+            isSideways: 0,
+            bidiLevel: 0,
+        };
+
+        let rendering_mode = font_face.get_recommended_rendering_mode_default_params(
+            em_size,
+            self.device_pixel_ratio,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+        );
 
-        let vmetrics = font.metrics();
-        let scale = (size.as_f32_pts() * self.device_pixel_ratio * (96.0 / 72.0))
-            / f32::from(vmetrics.designUnitsPerEm);
+        let glyph_analysis = GlyphRunAnalysis::create(
+            &glyph_run,
+            self.device_pixel_ratio,
+            None,
+            rendering_mode,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+            0.0,
+            0.0,
+        )
+        .or_else(|_| Err(Error::MissingGlyph(c)))?;
```

Comment:
Is this error really correct? We shouldn't be returning a missing glyph error if that's not the actual problem.

Same goes for all the other errors in this function following this one.

Reply:
I've gone for adding a `DirectWriteError` variant which stores the `HRESULT` error code returned from these functions. We could potentially give the variant a more specific name like `GlyphRasterizationFailed`?